node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -887,165 +880,355 @@ async fn process_incoming_peer_message<Context>(
+	RelayParentUnknown,
+	/// Peer is not present in the subsystem state.
+	UnknownPeer,
+	/// Peer has not declared its para id.
+	UndeclaredCollator,
+	/// We're assigned to a different para at the given relay parent.
+	InvalidAssignment,
+	/// Collator is trying to build on top of occupied core
+	/// when async backing is disabled.
+	CoreOccupied,
+	/// An advertisement format doesn't match the relay parent.
+	ProtocolMismatch,
+	/// Para reached a limit of seconded candidates for this relay parent.
+	SecondedLimitReached,
+	/// Advertisement is invalid.
+	Invalid(InsertAdvertisementError),
+	/// Failed to query prospective parachains subsystem.
+	ProspectiveParachainsUnavailable,
+}
+
+impl AdvertisementError {
+	fn reputation_changes(&self) -> Option<Rep> {
+		use AdvertisementError::*;
+		match self {
+			InvalidAssignment => Some(COST_WRONG_PARA),
+			RelayParentUnknown | UndeclaredCollator | CoreOccupied | Invalid(_) =>
+				Some(COST_UNEXPECTED_MESSAGE),
+			UnknownPeer |
+			ProtocolMismatch |
+			SecondedLimitReached |
+			ProspectiveParachainsUnavailable => None,
+		}
+	}
+}
+
+async fn handle_advertisement<Sender>(
+	sender: &mut Sender,
+	state: &mut State,
+	relay_parent: Hash,
+	peer_id: &PeerId,
+	prospective_candidate: Option<(CandidateHash, Hash)>,
+) -> std::result::Result<(), AdvertisementError>
+where
+	Sender: CollatorProtocolSenderTrait,
+{
+	let _span = state
+		.span_per_relay_parent
+		.get(&relay_parent)
+		.map(|s| s.child("advertise-collation"));
+
+	let per_relay_parent = state
+		.per_relay_parent
+		.get_mut(&relay_parent)
+		.ok_or(AdvertisementError::RelayParentUnknown)?;
+
+	let relay_parent_mode = per_relay_parent.prospective_parachains_mode;
+	let assignment = per_relay_parent.assignment;
+
+	let peer_data = state.peer_data.get_mut(&peer_id).ok_or(AdvertisementError::UnknownPeer)?;
+	let para_id = peer_data.collating_para().ok_or(AdvertisementError::UndeclaredCollator)?;
+
+	let core_state = match assignment.current {
+		Some((id, core_state)) if id == para_id => core_state,
+		_ => return Err(AdvertisementError::InvalidAssignment),
+	};
+
+	if !relay_parent_mode.is_enabled() && core_state.is_occupied() {
+		return Err(AdvertisementError::CoreOccupied)
+	}
+
+	// TODO: only fetch a collation if it's built on top of backed nodes in fragment tree.
+	// https://github.com/paritytech/polkadot/issues/5923
+	let is_seconding_allowed = match (relay_parent_mode, prospective_candidate) {
+		(ProspectiveParachainsMode::Disabled, _) => true,
+		(ProspectiveParachainsMode::Enabled, Some((candidate_hash, parent_head_data_hash))) => {
+			let active_leaves = state.active_leaves.keys().copied();
+			is_seconding_allowed(
+				sender,
+				relay_parent,
+				candidate_hash,
+				parent_head_data_hash,
+				para_id,
+				active_leaves,
+			)
+			.await
+			.ok_or(AdvertisementError::ProspectiveParachainsUnavailable)?
+		},
+		_ => return Err(AdvertisementError::ProtocolMismatch),
+	};
+
+	if !is_seconding_allowed {
```

Comment:
From [#5054](https://github.com/paritytech/polkadot/issues/5054): 

we need to make sure we _accept_ the advertisement (those are already throttled by `MAX_DEPTH` + peer-set size) but not request it until that point.

Reply:
We can defer this until #5923 but it is important to get this right because otherwise collators can spam us with unbounded numbers of advertisements.

Reply:
It doesn't actually resolves the spam issue. With the current code the number of collations is limited by MAX_DEPTH + 1 collations per parent per collator and by MAX_DEPTH + 1 of seconded collations.

Even if we add filtering based on the parent in a fragment tree, collators still can submit for example MAX_DEPTH + 1 candidates based on the same para head (most recent for this relay parent even) and it would be valid.

Reply:
My point is that if we reject advertisements without storing them somewhere the remote node can send us an infinite number of advertisements.

Reply:
Given a todo it never rejects advertisements (as of now), if `is_seconding_allowed` returns false we should return before requesting + put this advertisement in some special queue, but I deferred it to a different issue because it requires significant changes.

Reply:
I am fine with deferring this but I am concerned that the intent of the TODO will be lost and we will introduce a bug.

Reply:
No worries I do understand the desired behaviour